### ISSUE TITLE
NaN and Infinity can be produced by multiplication

### ISSUE THREAD

**Issue by qiemem** [NetLogo/NetLogo#895]
`2 ^ 1024` errors correctly, but:

```
observer> show (2 ^ 1023) * 2
observer: Infinity
observer> show ((2 ^ 1023) * 2) / ((2 ^ 1023) * 2)
observer: NaN
```

It looks like `_mult` doesn't call `isValidDouble`. That said, we should just use Java 8's new `isFinite` instead. We definitely need to thoroughly benchmark any of these changes since they touch basic math operations.

**Comment by SethTisue** [comment]
`validDouble` should be changed to call `isFinite`, you mean. `validDouble` doesn't return a boolean, it actually errors if the input isn't valid

**Comment by SethTisue** [comment]
there are a bunch of other prims (including the bytecode-generated version of `_plus`) that ought to call `validDouble` or `newValidDouble` but don't.

**Comment by qiemem** [comment]
Ah, yes, you're right.

**Comment by mrerrormessage** [comment]
The fix here is to ensure that all prims which generate a new number call `validDouble` or `newValidDouble` on their result before returning it. The following prims are the ones which can return a number:

* `count`
* `countother`
* `countotherwith`
* `countwith`
* `minus`
* `nsum`
* `nsum4`
* `patchvariabledouble`
* `plus`
* `random`
* `randomconst`
* `sum`
* `turtlevariabledouble`
* `unaryminus`
* `randomorrandomfloat`
* `abs`
* `acos`
* `approximatehsb`
* `approximatehsbold`
* `approximatergb`
* `asin`
* `atan`
* `behaviorspacerunnumber`
* `ceil`
* `cos`
* `distance`
* `distancenowrap`
* `distancexy`
* `distancexynowrap`
* `div`
* `dx`
* `dy`
* `exp`
* `floor`
* `int`
* `length`
* `linkheading`
* `linklength`
* `ln`
* `log`
* `max`
* `maxpxcor`
* `maxpycor`
* `mean`
* `median`
* `min`
* `minpxcor`
* `minpycor`
* `mod`
* `monitorprecision`
* `mult`
* `nanotime`
* `newseed`
* `patchsize`
* `position`
* `pow`
* `precision`
* `processors`
* `randomexponential`
* `randomfloat`
* `randomgamma`
* `randomnormal`
* `randompoisson`
* `randompxcor`
* `randompycor`
* `readfromstring`
* `remainder`
* `round`
* `scalecolor`
* `sin`
* `sqrt`
* `standarddeviation`
* `subtractheadings`
* `tan`
* `ticks`
* `timer`
* `towards`
* `towardsnowrap`
* `towardsxy`
* `towardsxynowrap`
* `variance`
* `worldheight`
* `worldwidth`
* `wrapcolor`
* `fileread`
* `mousexcor`
* `mouseycor`
* `hubnetinqsize`
* `hubnetoutqsize`
* `plotxmax`
* `plotxmin`
* `plotymax`
* `plotymin`
* `distancexyz`
* `distancexyznowrap`
* `dz`
* `linkpitch`
* `maxpzcor`
* `minpzcor`
* `oheading`
* `opitch`
* `oroll`
* `oxcor`
* `oycor`
* `ozcor`
* `randompzcor`
* `randomzcor`
* `towardspitch`
* `towardspitchnowrap`
* `towardspitchxyz`
* `towardspitchxyznowrap`
* `worlddepth`

Obviously, some of these primitives don't need to be checked. We can safely say that `count` will never return a `NaN`. Nonetheless, most of them *do* need to be checked in both their bytecode generated forms (where applicable) and their generated forms.

**Comment by mrerrormessage** [comment]
Fixed in #970 . I haven't noticed any detrimental performance affects since this change was merged, but if there have been any they could be tackled by adding checks at the edge of double-expressions. I'm not opening an issue for that at the moment, but may do so in the future.
